core/tests/performance_framework/perf_services.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2016 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Contains a utility for fetching performance data using Selenium and
+Browsermob-proxy.
+"""
+
+import os
+import time
+import urlparse
+from browsermobproxy import Server
+from selenium import webdriver
+
+# Used for headless testing i.e, making browser windows invisible to the user.
+from xvfbwrapper import Xvfb
+
+from core.tests.performance_framework import perf_domain
+
+CHROME_DRIVER_PATH = os.path.join(
+    '..', 'node_modules', 'protractor', 'selenium', 'chromedriver_2.21')
+
+BROWSERMOB_PROXY_PATH = os.path.join(
+    '..', 'oppia_tools', 'browsermob-proxy-2.1.1', 'bin', 'browsermob-proxy')
+
+
+class SeleniumPerformanceDataFetcher(object):
+    """Fetches performance data for locally served Oppia pages using Selenium
+    and Browsermob-proxy.
+
+    Selenium is used to programitically interact with a browser.
+    Browsermob-proxy is used to capture HTTP Archive (referred to as HAR) data.
+
+    The HTTP Archive format is a JSON-formatted archive file format used for
+    logging a web browser's interaction with a site. It contains detailed
+    performance data, including information about page loading and displaying
+    and per-resource statistics. Each entry contains the URL requested and
+    request and response headers. For additional details, please see:
+    https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/HAR/Overview.html
+
+    The HTTP Archive data is retrieved using the proxy server by recording the
+    communication between the client and server.
+    """
+
+    # Chrome is the default browser that selenium will use for extracting
+    # performance stats.
+    DEFAULT_BROWSER_SOURCE = 'chrome'
+    SUPPORTED_BROWSER_SOURCES = ['chrome', 'firefox']
+
+    # Duration (in seconds) to wait for the complete page to load, otherwise
+    # XHR requests made post initial page load will not be recorded.
+    WAIT_DURATION_SECS = 3
+
+    def __init__(self, browser=DEFAULT_BROWSER_SOURCE):
+        if browser in self.SUPPORTED_BROWSER_SOURCES:
+            self.browser = browser
+        else:
+            error_msg = 'Unsupported browser specified: %s' % browser
+            raise ValueError(error_msg)
+
+        self.server = self.driver = self.proxy = None
+
+    def get_page_metrics_for_url(self, page_url):
+        """Returns a PageSessionMetrics domain object initialized using
```

Comment:
I really am strongly opposed to referencing HAR in either of these two methods' documentation. It's an implementation detail that will not affect the developer using these two methods, so there's not a valuable reason for the developer to know about HAR files here. HAR files are described in the docstring of this class which is already more than I would normally expect, but it seems reasonable in the sense that this fetcher does use HAR internally and may eventually be extended to support the developer passing in a HAR-formatted dict.

These two methods don't actually need to be implemented using HAR files underneath in order to still function correctly (in the eyes of the programmer using the methods). For that reason, ``get_page_metrics_for_url`` and ``get_page_metrics_from_cached_state`` do not need to have HAR in their documentation.